src/mlpack/tests/ann_layer_test.cpp

```diff
@@ -3043,6 +3043,89 @@ TEST_CASE("AtrousConvolutionLayerPaddingTest", "[ANNLayerTest]")
   module2.Backward(input, output, delta);
 }
 
+/**
+ * Tests the GroupNorm layer.
+ */
+TEST_CASE("GroupNormTest", "[ANNLayerTest]")
+{
+  arma::mat input, output, backwardOutput;
+  input = {
+    { 2, 0, 1 },
+    { 3, 1, 2 },
+    { 5, 1, 3 },
+    { 7, 2, 4 },
+    { 11, 3, 5 },
+    { 13, 5, 6 },
+    { 17, 8, 7 },
+    { 19, 13, 8 }
+  };
+
+  GroupNorm<> model(2, 4, 2);
+  model.Reset();
+
+  model.Forward(input, output);
+  arma::mat result;
+  result = {
+    { -1.1717001972, -1.4142135482, -1.3416407811 },
+    { -0.6509445540, 0.0000000000 , -0.4472135937 },
+    { 0.3905667324 , 0.0000000000 , 0.4472135937  },
+    { 1.4320780188 , 1.4142135482 , 1.341640781   },
+    { -1.2649110634, -1.1283296293, -1.3416407811 },
+    { -0.6324555317, -0.5973509802, -0.4472135937 },
+    { 0.6324555317 , 0.1991169934 , 0.4472135937  },
+    { 1.2649110634 , 1.5265636161 , 1.3416407811  }
+  };
+
+  CheckMatrices(output, result, 1e-5);
+
+  model.Backward(output, output, backwardOutput);
+  std::cout.precision(20);
+  backwardOutput.raw_print(std::cout);
+}
+
+/**
+ * GroupNorm layer numerical gradient test.
+ */
+TEST_CASE("GradientGroupNormTest", "[ANNLayerTest]")
+{
+  // Add function gradient instantiation.
+  struct GradientFunction
+  {
+    GradientFunction() :
+        input(arma::randn(10, 256)),
+        target(arma::zeros(1, 256))
+    {
+      model = new FFN<NegativeLogLikelihood<>, NguyenWidrowInitialization>();
+      model->Predictors() = input;
+      model->Responses() = target;
+      model->Add<IdentityLayer<> >();
+      model->Add<Linear<> >(10, 10);
+      model->Add<GroupNorm<> >(2, 5, 5);
+      model->Add<Linear<> >(10, 2);
+      model->Add<LogSoftMax<> >();
+    }
+
+    ~GradientFunction()
+    {
+      delete model;
+    }
+
+    double Gradient(arma::mat& gradient) const
+    {
+      double error = model->Evaluate(model->Parameters(), 0, 256, false);
+      model->Gradient(model->Parameters(), 0, gradient, 256);
+      return error;
+    }
+
+    arma::mat& Parameters() { return model->Parameters(); }
+
+    FFN<NegativeLogLikelihood<>, NguyenWidrowInitialization>* model;
+    arma::mat input, target;
+  } function;
+
+  REQUIRE(CheckGradient(function) <= 1e-4);
```

Comment:
Looks like the gradient explodes after a couple of epochs. We could add gradient clipping, but that would contradict what we are trying to test here. I tested it with a lower batch-size as well, but that doesn't solve the problem. So maybe instead of testing adding a numerical gradient check, what about we just use the `GroupNorm` layer with predefined `gamma` and `beta` and compare that against some predefined values, we e.g. get from PyTorch.

Reply:
But its not reaching tests. It fails in build itself

Reply:
```GradientGroupNormTest
-------------------------------------------------------------------------------
/home/vsts/work/1/s/src/mlpack/tests/ann_layer_test.cpp:3089
...............................................................................

/home/vsts/work/1/s/src/mlpack/tests/ann_layer_test.cpp:3126: FAILED:
  REQUIRE( CheckGradient(function) <= 1e-4 )
with expansion:
  nan <= 0.0001
```

is what I see on the latest azure build.

Reply:
Yes,  I see. For some reason the gradients keeps on increasing. Not  sure why this is happening. The forward and backward pass  output  matches with the pytorch implementation.
Is there any way to check the gradient from pytorch.

Reply:
https://discuss.pytorch.org/t/how-to-print-the-computed-gradient-values-for-a-network/34179/2 has some more pointers regarding getting the gradient.

Reply:
Nice. I will look into it. The last commit made the group norm behave like layer norm by making group count = total channels , let's see if it passes or not.